File: framework/src/play/src/main/scala/play/api/http/websocket/Message.scala
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <https://www.lightbend.com>
 */
package play.api.http.websocket

import akka.util.ByteString

/**
 * A WebSocket message.
 *
 * This is a high level API intended for common simple use cases.  It allows handling and sending of full
 * WebSocket messages, as well as close and ping/pong messages.  It will buffer fragmented messages up until
 * a limit, and does not allow streaming in fragments.
 */
sealed trait Message

/**
 * A text message.
 *
 * @param data The data of the text message.
 */
case class TextMessage(data: String) extends Message

/**
 * A binary message.
 *
 * @param data The data of the binary message.
 */
case class BinaryMessage(data: ByteString) extends Message

/**
 * A close message.
 *
 * @param statusCode The close status code.
 * @param reason The reason it was closed.
 */
case class CloseMessage(statusCode: Option[Int] = Some(CloseCodes.Regular), reason: String = "") extends Message

object CloseMessage {
  def apply(statusCode: Int): CloseMessage =
    CloseMessage(Some(statusCode), "")
  def apply(statusCode: Int, reason: String): CloseMessage =
    CloseMessage(Some(statusCode), reason)
}

/**
 * A ping message.
 *
 * @param data The application data.
 */
case class PingMessage(data: ByteString) extends Message

/**
 * A pong message.
 *
 * @param data The application data.
 */
case class PongMessage(data: ByteString) extends Message

/**
 * An exception that, if thrown by a WebSocket source, will cause the WebSocket to be closed with the given close
 * message. This is a convenience that allows the WebSocket to close with a particular close code without having
 * to produce generic Messages.
 */
case class WebSocketCloseException(message: CloseMessage) extends RuntimeException(message.reason, null, false, false)

File: framework/src/play/src/main/java/play/http/websocket/Message.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.http.websocket;

import akka.util.ByteString;

import java.util.Optional;

/**
 * A WebSocket message.
 */
public abstract class Message {

    // private constructor to seal it
    private Message() {
    }

    /**
     * A text WebSocket message
     */
    public static class Text extends Message {
        private final String data;

        public Text(String data) {
            this.data = data;
        }

        public String data() {
            return data;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Text text = (Text) o;

            return data.equals(text.data);

        }

        @Override
        public int hashCode() {
            return data.hashCode();
        }

        @Override
        public String toString() {
            return "TextWebSocketMessage('" + data + "')";
        }
    }

    /**
     * A binary WebSocket message
     */
    public static class Binary extends Message {
        private final ByteString data;

        public Binary(ByteString data) {
            this.data = data;
        }

        public ByteString data() {
            return data;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Binary binary = (Binary) o;

            return data.equals(binary.data);

        }

        @Override
        public int hashCode() {
            return data.hashCode();
        }

        @Override
        public String toString() {
            return "BinaryWebSocketMessage('" + data + "')";
        }
    }

    /**
     * A ping WebSocket message
     */
    public static class Ping extends Message {
        private final ByteString data;

        public Ping(ByteString data) {
            this.data = data;
        }

        public ByteString data() {
            return data;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Ping ping = (Ping) o;

            return data.equals(ping.data);

        }

        @Override
        public int hashCode() {
            return data.hashCode();
        }

        @Override
        public String toString() {
            return "PingWebSocketMessage('" + data + "')";
        }
    }

    /**
     * A pong WebSocket message
     */
    public static class Pong extends Message {
        private final ByteString data;

        public Pong(ByteString data) {
            this.data = data;
        }

        public ByteString data() {
            return data;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Pong pong = (Pong) o;

            return data.equals(pong.data);

        }

        @Override
        public int hashCode() {
            return data.hashCode();
        }

        @Override
        public String toString() {
            return "PongWebSocketMessage('" + data + "')";
        }
    }

    /**
     * A close WebSocket message
     */
    public static class Close extends Message {
        private final Optional<Integer> statusCode;
        private final String reason;

        public Close(int statusCode) {
            this(statusCode, "");
        }

        public Close(int statusCode, String reason) {
            this(Optional.of(statusCode), reason);
        }

        public Close(Optional<Integer> statusCode, String reason) {
            this.statusCode = statusCode;
            this.reason = reason;
        }

        public Optional<Integer> code() {
            return statusCode;
        }

        public String reason() {
            return reason;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Close close = (Close) o;

            return statusCode.equals(close.statusCode) && reason.equals(close.reason);
        }

        @Override
        public int hashCode() {
            int result = statusCode.hashCode();
            result = 31 * result + reason.hashCode();
            return result;
        }

        @Override
        public String toString() {
            return "CloseWebSocketMessage(" + statusCode + ", '" + reason + "')";
        }
    }


}
